src/Npgsql/SqlQueryParser.cs

```diff
@@ -33,23 +33,17 @@ class SqlQueryParser
             List<NpgsqlStatement> statements,
             bool standardConformingStrings,
             bool deriveParameters = false)
-            => ParseRawQuery(sql.AsSpan(), parameters, statements, standardConformingStrings, deriveParameters);
-
-        void ParseRawQuery(
-            ReadOnlySpan<char> sql,
-            NpgsqlParameterCollection parameters,
-            List<NpgsqlStatement> statements,
-            bool standardConformingStrings,
-            bool deriveParameters)
         {
             Debug.Assert(deriveParameters == false || parameters.Count == 0);
+            var span = sql.AsSpan();
 
-            NpgsqlStatement statement = null!;
+            var positionalParameters = parameters.Count > 0 && parameters[0].ParameterName.Length == 0;
```

Comment:
> In #3693 I can probably improve positional support.

In my defense, I procrastinated on reviewing #3693 partially because I wanted to move ahead with positional parameters here first :rofl: Yeah, the two are definitely connected.

> I'd like it if we have a cheap check for positional

The above already seems pretty cheap :) 

I actually tried to do something like this, but it turns out that for SqlQueryParser we're in positional mode *only* if there's a positional parameter (and not if there are no parameters), whereas for NpgsqlCommand.AutoPrepare/Prepare we're in positional mode also if there aren't any parameters. So it's a bit tricky.

> an exception when param styles are mixed, which is something I'm missing here.

Good catch - I'll add that to NpgsqlCommand.ValidateParameters.

Reply:
>The above already seems pretty cheap :)

Well the cheap part would include being cheap to validate mixed inputs

Reply:
Take a look at how I've implemented validation in the latest commit. The idea is that we're already going over parameters in ValidateParameter (for other reasons), so I'm trying to pile anything that needs to look at the parameters in there. This includes calculating whether we have output parameters, and also whether we're mixing parameter styles.